Remove `context` from model evaluation (use `model.context` instead)

penelopeysm opened the pull request:
## Summary

This PR modifies the _model evaluation function_, `model.f` for some `model::DynamicPPL.Model`, to not take a context as its third argument. Thus, its signature looks like `f(model, varinfo, args...; kwargs...)` where `args` and `kwargs` are forwarded from the function that defines the model.

During model evaluation, the tilde-pipeline would always dispatch on the `__context__` argument, and would completely ignore the `__model__.context`. It has now been changed so that it dispatches on `__model__.context`.

As a result of this, we can remove the `context` argument from most model evaluation code as well as LogDensityFunction. This simplifies a lot of code.

There are a handful of minor follow-up points:

### Combination of `__context__` and `__model__.context`

Inside the _model evaluation function_, the model's context was previously being ignored. However, if one were to call `evaluate!!(model, varinfo, context)`, the model's context does not get ignored: it is combined together with `context` in order to form a larger context stack, which _then_ becomes the actual evaluation context.

https://github.com/TuringLang/DynamicPPL.jl/blob/3e54c2d8c5f790d8ba83061671e22783dadbc164/src/model.jl#L942-L944

After this PR, the intention is that if you want to keep that combining behaviour, you should do it manually. However, several 'outs' are provided:

1. `sample!!(rng, model, varinfo, sampler)` is provided. This wraps `model.context` in a `SamplingContext(rng, sampler)` before calling `evaluate!!`. This takes care of most invocations of `evaluate!!` where this combining behaviour was being used.

   (As a bonus, this also helps us remove the rng and sampler arguments from `evaluate!!`, which greatly simplifies that function: #720)

2. `evaluate!!(model, varinfo, context)` still exists, but is dep-warned.

3. `_evaluate!!(model, varinfo, context)` still exists with the same behaviour without a depwarn. This is only retained because submodels depend on this and I would like to leave this change for a subsequent PR, as that needs to be reasoned about more deeply.

### Models as callables

Previously, `(model::Model)(args...)` would forward to `evaluate!!(args...)`. I think this behaviour is quite dangerous (https://github.com/TuringLang/DynamicPPL.jl/pull/629) and even in that PR it was mentioned that we should be more explicit about what args we take, which I fully agree with.

Along with cleaning up `evaluate!!`, this PR also cleans up models as callables such that the only allowed signatures are `model()`, `model(rng)`, `model(varinfo)`, and `model(rng, varinfo)`. Thus you are no longer allowed to specify a context (you should `contextualize` the model instead), or a sampler (the default sampler was `SampleFromPrior()` and given that `model(...)` was always being used to sample from the prior, it seems _hugely_ unlikely that anybody was really passing a different sampler, and if they were, they can jolly well call `first(DynamicPPL.sample!!(rng, model, varinfo, sampler))`).

### Remaining test failure

The remaining test failure (Julia-pre) is unrelated to this PR.

Closes #720
Closes #951

yebai commented:
That's an excellent direction. When duplicate contexts were first introduced, I was always bothered by them. We merged the relevant PRs to avoid blocking incremental improvements.

yebai commented:
One minor comment on naming: `sample!!` will likely confuse people since it is ubiquitously used for sampling (posterior) distributions using MCMC. Perhaps consider an alternative, e.g., `evaluate_and_sample!!`?

penelopeysm commented:
> this runtime seems to have ~doubled

The time on this branch is the same as on [the breaking branch](https://github.com/TuringLang/DynamicPPL.jl/pull/901#issuecomment-2930330594), though: so maybe it came from an earlier merge into there? I'm just in the process of updating all those branches so can take a look in a while

mhauru commented:
My bad, I wasn't thinking when picking the thing to compare to. Nothing to do there then.

I'm still mid-review, but wondering about what to do about accumulators. I would like to say that I was hasty in merging the first accumulator stuff into `breaking`, because it's taking a while now to finish that stuff. It's releaseable in that tests pass, but I'm not sure it makes sense to release the accumulator stuff in parts. Thus we should move it to a different branch and clear the path for this to make it to a minor release. On the other hand though, this has been developed on top of accumulators, and I wouldn't want anyone to spend time disentangling the two with `cherry-pick` or `rebase`.

penelopeysm commented on src/extract_priors.jl:
```diff
@@ -116,7 +116,8 @@ function extract_priors(rng::Random.AbstractRNG, model::Model)
     # can't push new variables without knowing the num_produce. Remove this when possible.
     varinfo = setaccs!!(varinfo, (PriorDistributionAccumulator(), NumProduceAccumulator()))
-    varinfo = last(evaluate!!(model, varinfo, SamplingContext(rng)))
+    new_model = contextualize(model, SamplingContext(rng, model.context))
+    varinfo = last(evaluate!!(new_model, varinfo))
```
Hmmmm, chances are I did this before realising that it was a common pattern across the codebase and writing `sample!!`.

penelopeysm commented on src/model.jl:
```diff
@@ -794,15 +800,23 @@ julia> # Now `a.x` will be sampled.
+function (model::Model)(varinfo::AbstractVarInfo)
+    return model(Random.default_rng(), varinfo)
+end
+# ^ Weird Documenter.jl bug means that we have to write the two above separately
+# as it can only detect the `function`-less syntax.
```
There are a few issues about callable structs but I couldn't find one that was explicitly about this. I'll try to make an MWE

penelopeysm commented:
Okay, I fixed all the stuff. I renamed it to `evaluate_and_sample!!` too, but with a big warning in the changelog that this is liable to change (depending on what we do with SamplingContext). I think that should address the technical bits. The remaining question is whether we split this onto a separate branch. I'm inclined to not bother, partly because this isn't ready for release yet anyway (I need to fix it for submodels), and partly because of the hassle of cherry-picking (the simplification of leaf contexts helped a lot here, and I'm scared I might overlook correctness issues if I were to replay this on main).

I think the main argument for releasing it now is that I could use it upstream in some of the Turing sampler + LDF work, but I'm actually happy to hold off on that until we have a resolution to #955, and #955 in turn certainly needs to wait for accs to be done. So I think the happiest course of action is just for me to find other things to do until we are ready to release accs.
